app/models/potential_matcher/prompt_tag_type_info.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+# A class used for gathering information about the rough frequency of different
+# required tag types in all of the prompts in a collection. Used to figure out
+# what would be a good "index type" -- that is, a type that we can use to try
+# to look up matches with.
+class PromptTagTypeInfo
+  # We use a larger batch size here because prompts are smaller than signups
+  # (which can contain many prompts), so we can store more of them in memory at
+  # the same time.
+  def initialize(collection, batch_size = 300)
+    @collection = collection
+    @batch_size = batch_size
+
+    load_collection_info
+    initialize_data_tables
+  end
+
+  private
+
+  # Load information about the passed-in collection.
+  def load_collection_info
+    @total_prompts = @collection.prompts.count
+    @total_signups = @collection.signups.count
+    @settings = @collection.challenge.potential_match_settings
+    @required_types = @settings.required_types
+
+    @limit_first_tags = [10, @total_signups / 10].min
+    @limit_prompts_with_any = [10, @total_prompts / 10].max
```

Comment:
A note here, move the numbers here into separate constants in config.yml, which you can then access here with "ArchiveConfig.[constant name]" eg ArchiveConfig.FIRST_TAGS_LIMIT etc. -- this will go into cfengine and can then be tweaked live if we find a better number